### DISCUSSION TITLE
How to migrate from bittwarden_rs (Version: 1.11.0-9ff577a7) to latest vaultwarden version

### DISCUSSION THREAD

**Discussion by 404erreur** [dani-garcia/vaultwarden#3095] [Q&A] [answered]
Hello,

Do you know if an update documentation exist for this update ?

bittwarden_rs (Version: 1.11.0-9ff577a7) to latest vaultwarden version

I don't use docker, it a installation on system directly.

Many thanks in advance.

Regards
Matthias

**Comment by BlackDex** [comment]
Check out the big pinned topic https://github.com/dani-garcia/vaultwarden/discussions/1642

**Comment by 404erreur** [comment]
Hello,

Many thanks for your reply.
I have already consult this post.

But it's not very clear for me.

I'm currently used bitwarden_RS Version 2.9.0  with v2.10.0.patch and v2.11.0.patch.
In the admin panel I can check Version: 1.11.0-9ff577a7.

After the project rename, (sorry if I'm confused), but some element seem to be now separated.
api/web_vault.

There is an all in one package, with an procedure like:

- Update the sofware requirement to the version xxxx (exemple nodejs, etc ...)
- Untar this archive
- Replace this older folder xxxxx by the new folder name yyyyyy.
- Execute this script for upgrading database table etc .....

Many thanks in advance for your time.

I guess this explaination can help lot of user :)

ps: perhaps I also need to make more than one upgrade ? exemple ugrading to 2.21 version and then upgrading 2.5 version.

Best regards
Matthias

**Reply by BlackDex** [reply]
I'm not sure what is unclear there.
Maybe if you provide how you installed it we could give a better answer.
But in theory, if you used docker containers, it is as simple as it is stated there, rename `bitwardenrs` to `vaultwarden` pull en re-create the container.

**Comment by 404erreur** [comment]
Hello,

Many thanks for your reply.

I don't use docker
So I need to know witch folder I need to replace.


### Versions information:

I'm currently used bitwarden_RS Version 2.9.0 with v2.10.0.patch and v2.11.0.patch.
In the admin panel I can check Version: 1.11.0-9ff577a7.




### Directory installation:

```
ls /opt/bitwarden/
bitwarden_rs  bitwarden_rs.d  build  data  deps  examples  incremental  web-vault
```


### Environemnt Variables:

```
cat /etc/bitwarden_rs.env
SIGNUPS_ALLOWED=false
WEBSOCKET_ENABLED=true
ADMIN_TOKEN=XXXXXX
ROCKET_ADDRESS=127.0.0.1
WEBSOCKET_ADDRESS=127.0.0.1
SMTP_HOST=127.0.0.1
SMTP_FROM=XXXXXX
SMTP_PORT=25
SMTP_SSL=false
ENABLE_DB_WAL='false'
DATABASE_URL="mysql://bitwarden:XXXXXX@127.0.0.1:3306/bitwarden"
```



### Init Script:

```
# Bitwarden systemd service file

[Unit]
Description=Bitwarden Server (Rust Edition)
Documentation=https://github.com/dani-garcia/bitwarden_rs
After=network.target

[Install]
WantedBy=multi-user.target

[Service]
Type=simple
User=www-data
Group=www-data
PermissionsStartOnly=true
EnvironmentFile=/etc/bitwarden_rs.env
ExecStart=/opt/bitwarden/bitwarden_rs
Restart=on-failure
# Isolate bitwarden_rs from the rest of the system
#PrivateTmp=true
#PrivateDevices=true
#ProtectHome=true
#ProtectSystem=strict
# Only allow writes to the following directory and set it to the working directory (user and password data are stored here)
WorkingDirectory=/opt/bitwarden/
ReadWriteDirectories=/opt/bitwarden/
```

Another possible ways for me can be install fresh installation of vaulwarden, and copy the current data.
But I don't exactly know what I need to take?

I suppose the database and the env file, but which folder I need to copy ? 

```
/opt/bitwarden/data
```

And only this one ?

Regards
Matthias

**Reply by BlackDex** [reply]
Just backup the data. And install Vaultwarden and point out to the exact same folder. Or copy the whole bitwarden folder to a vaultwarden folder and point it there. And have everything exactly the same, that should work.

**Reply by 404erreur** [reply]
Many thanks, 

I 'll try asap (I hope next week)

I'll give a feedback and pay a beer if all is working great :)

Regards
Matthias

**Comment by 404erreur** [comment, marked as answer]
Hello all,

Sorry for the time of my feedback.
The migration is done and all work great.

The only one problem was on the database upgrade on mariadb 10.6, but I use a workaround. (migrating the database on the older mysql 5.7 server,  and then export/import to mariadb 10.6.

The method apply for me:

Regards
Matthias


--------------------


# Vaultwarden

How to compile and install Vaultwarden
(https://bitwarden.com/) in [Rust](https://www.rust-lang.org/). 


## Compilation

### Dependencies installation :

```bash
sudo apt install pkg-config libssl-dev build-essential default-libmysqlclient-dev git
```


### Rust installation

`rustup` installation, this one can install Rust compiler:

```bash
curl https://sh.rustup.rs -sSf > rustup.sh
chmod +x rustup.sh
```


Rust Compiler installation :

```bash
./rustup.sh --default-host x86_64-unknown-linux-gnu --default-toolchain nightly
```

Sourcing the compiler environment file :

```bash
source $HOME/.cargo/env
```


### Updating Rust (if necessary)

```bash
rustup update
```


### Compile

Cloning vaultwarden project:

```bash
git clone https://github.com/dani-garcia/vaultwarden
```


`Vaultwarden` compilation:

```bash
cd vaultwarden
cargo build --release --features mysql
cd -

# features can be sqlite, mysql, or postgresql
```

The result will be in `vaultwarden/target/release/` folder.

### Web interface inclusion:

```bash
wget https://github.com/dani-garcia/bw_web_builds/releases/download/v2022.12.0/bw_web_v2022.12.0.tar.gz
tar xvf bw\_web\_v2022.12.0.tar.gz
```

For an update, deleting the previous web interface:

```bash
rm -rf vaultwarden/target/release/web-vault/
```


Then, moving it into the builded release :

```bash
mv web-vault/ vaultwarden/target/release/web-vault/
```


## Installation

In this exemple, We will install `Vaultwarden` in  `/opt/vaultwarden` directory with `www-data` rights :

```bash
rsync -a --info=progress2 vaultwarden/target/release/ /opt/vaultwarden/
chown -R www-data: /opt/vaultwarden
```


Setup the `systemd` init file `/etc/systemd/system/vaultwarden.service` :

```
[Unit]
Description=Vaultwarden Server (Rust Edition)
Documentation=https://github.com/dani-garcia/vaultwarden
After=network.target

[Service]
# The user/group vaultwarden is run under. the working directory (see below) should allow write and read access to this user/group
User=www-data
Group=www-data
# The location of the .env file for configuration
EnvironmentFile=/etc/vaultwarden.env
# The location of the compiled binary
ExecStart=/opt/vaultwarden/vaultwarden
# Set reasonable connection and process limits
LimitNOFILE=1048576

# Isolate vaultwarden from the rest of the system
PrivateTmp=true
PrivateDevices=true
ProtectHome=true
ProtectSystem=strict
# Only allow writes to the following directory and set it to the working directory (user and password data are stored here)
WorkingDirectory=/opt/vaultwarden/
ReadWriteDirectories=/opt/vaultwarden/

[Install]
WantedBy=multi-user.target
```

Admin token creation :

```bash
openssl rand -base64 48
```

Environments variable file `/etc/vaultwarden.env` :

```
SIGNUPS_ALLOWED=false
WEBSOCKET_ENABLED=true
ADMIN_TOKEN=${PREVIOUS_TOKEN_CREATED}
ROCKET_ADDRESS=127.0.0.1
WEBSOCKET_ADDRESS=127.0.0.1
SMTP_HOST=127.0.0.1
SMTP_FROM=vaultwarden@example.org
SMTP_PORT=25
SMTP_SSL=false
```

(https://github.com/dani-garcia/vaultwarden/blob/master/.env.template) for the documentation

Then:

```bash
sudo systemctl daemon-reload
systemctl enable --now vaultwarden
systemctl status vaultwarden
```


## Nginx

Nginx installation :

```bash
apt install nginx
```

Vhost content:

```nginx
server {
    listen 80;
    listen [::]:80;
    listen 443 ssl http2;
    listen [::]:443 ssl http2;
    server_name vaultwarden.example.org;

    access_log /var/log/nginx/vaultwarden.access.log;
    error_log /var/log/nginx/vaultwarden.error.log;

    ssl_certificate /etc/nginx/ssl/certtificat.crt;
    ssl_certificate_key /etc/nginx/ssl/certificat.key;

    ssl_session_timeout 5m;
    ssl_session_cache shared:SSL:5m;

    ssl_prefer_server_ciphers On;
    ssl_protocols TLSv1.2;
    ssl_ciphers 'EECDH+aRSA+AESGCM:EECDH+aRSA+SHA384:EECDH+aRSA+SHA256:EECDH:+CAMELLIA256:+AES256:+CAMELLIA128:+AES128:+SSLv3:!aNULL:!eNULL:!LOW:!3DES:!MD5:!EXP:!PSK:!DSS:!RC4:!SEED:!ECDSA';

    ssl_dhparam /etc/nginx/ssl/private/dhparam4096.pem;
    add_header Strict-Transport-Security max-age=15768000; # six months
    gzip off;

    if ($https != 'on') {
        rewrite ^/(.*)$ https://vaultwarden.example.org/$1 permanent;
    }

    root /var/www/html;

    # Allow large attachments
    client_max_body_size 128M;

    location ^~ '/.well-known/acme-challenge' {
        default_type "text/plain";
        root /var/www/certbot;
    }

    location / {
        include /etc/nginx/proxy_params;
        proxy_pass http://127.0.0.1:8000;
    }

    location /notifications/hub {
        proxy_http_version 1.1;
        include /etc/nginx/proxy_params;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_pass http://127.0.0.1:3012;
    }

    location /notifications/hub/negotiate {
        include /etc/nginx/proxy_params;
        proxy_pass http://127.0.0.1:8000;
    }
}
```

Creating DHparam file`/etc/nginx/ssl/dhparam4096.pem` :

```bash
openssl dhparam -out /etc/ssl/private/dhparam4096.pem 4096
```

Reload nginx :

```bash
nginx -t && nginx -s reload
```

## Migrating data

On the new installation

- replace data folder with this one in the previous installation
- database content
- environment file (check if the content need to be update)

**Reply by BlackDex** [reply]
Thanks for this great and nice explanation!
This could be useful for others.
